pkg/datasources/functions.go

```diff
@@ -60,43 +62,58 @@ var functionsSchema = map[string]*schema.Schema{
 
 func Functions() *schema.Resource {
 	return &schema.Resource{
-		Read:   ReadFunctions,
-		Schema: functionsSchema,
+		ReadContext: ReadContextFunctions,
+		Schema:      functionsSchema,
 	}
 }
 
-// todo: fix this. ListUserFunctions isn't using the right struct right now and also the signature of this doesn't support all the features it could for example, database and schema should be optional, and you could also list by account.
-func ReadFunctions(d *schema.ResourceData, meta interface{}) error {
+func ReadContextFunctions(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
 	databaseName := d.Get("database").(string)
 	schemaName := d.Get("schema").(string)
 
-	d.SetId("functions")
-	currentFunctions, err := snowflake.ListUserFunctions(databaseName, schemaName, db)
-	if err != nil {
-		log.Printf("[DEBUG] error listing functions: %v", err)
-		return nil
+	request := sdk.NewShowFunctionRequest()
+	if databaseName != "" {
```

Comment:
in this case no, since database and schema are both required. but we should fix the data source to not require these. this can come later. cleaned up logic in meantime